runtime/main/src/main/java/org/qbicc/runtime/main/VMHelpers.java

```diff
@@ -111,48 +104,71 @@ private static boolean isAssignableToLeaf(type_id valueTypeId, type_id toTypeId)
     private static void omError(c_int nativeErrorCode) throws IllegalMonitorStateException {
         int errorCode = nativeErrorCode.intValue();
         if (0 != errorCode) {
-            throw new IllegalMonitorStateException("error code is: " + errorCode);
+//            putchar('O');
+//            putchar('M');
+//            putchar('E');
+//            putchar('R');
+//            putchar('R');
+//            putchar(':');
+//            printInt(errorCode);
+            throw new IllegalMonitorStateException("error code: " + errorCode);
         }
     }
 
     // TODO: mark this with a "NoInline" annotation
     static void monitor_enter(Object object) throws IllegalMonitorStateException {
-        /* TODO: deal with racy nature of this creation */
-        if (objectMonitorNatives == null) {
-            objectMonitorNatives = new ConcurrentHashMap<>();
-        }
-
-        // TODO malloc(sizeof(class)) resulted in "invalid coercion of s64 to u64" this is a workaround
-        Stddef.size_t mutexAttrSize = sizeof(pthread_mutexattr_t.class);
-        ptr<?> attrVoid = malloc(word(mutexAttrSize.longValue()));
-        if (attrVoid.isNull()) {
-            throw new OutOfMemoryError(/*"Allocation failed"*/);
+        if (object == null) {
+            /* TODO skip for now. Object should never be null except that
+                classof_from_typeid is not currently implemented. */
+            return;
         }
-        ptr<pthread_mutexattr_t> attr = (ptr<pthread_mutexattr_t>)castPtr(attrVoid, pthread_mutexattr_t.class);
+        while (true) { /* facilitate restart if atomic set operation fails */
```

Comment:
We don't need to restart here if the CAS fails.  The usual pattern is:
```
previousValue = CAS(location, oldValue, newValue);
if (previousValue != oldValue) {
  // someone else installed a different value for oldValue
  newValue = previousValue;
}
use newValue as the result
```
Different encodings of CAS have different semantics on how they return the result but we just need a way to tell if the mutex slot has transitioned from null to something as the something should stick in that slot for the life of the object.  (Though later we can look at deflating and removing monitors as well)